### PR TITLE
More helpful note message when confusing a field with a method

### LINKED ISSUE #2392: Helpful error message when confusing a field with a method
I forget if there's already a bug on this, but it would be nice if, when you wrote ```A.B``` and ```A``` doesn't have a field named ```B```, but does have a nullary method named ```B```, the compiler gave a hint like "did you mean to write ```A.B()```?" It could also do vice versa (if you write A.B()``` and ```A``` is a class with field ```B```, it could ask "did you mean ```A.B```?)

I ran into this in trans, where I have frequently written ```bcx.ccx``` instead of ```bcx.ccx()```.

### PR DESCRIPTION
**Pull request by Nashenas88**
This fixes #2392

I'd like to thank @eddyb for helping me on this one! I wouldn't have gotten the complicated FnOnce check done without his help.

### CODE REVIEW

File: src/librustc_typeck/check/method/suggest.rs [hunk 1]
```diff
@@ -61,12 +64,44 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
...
+                        Ok(expr_string) => expr_string,
+                        _ => "s".into() // Default to a generic placeholder for the
+                                        // expression when we can't generate a string
+                                        // snippet
+                    };
+
+                    // Determine if the field can be used as a function in some way
+                    let fn_once_trait_did = match cx.lang_items.require(FnOnceTraitLangItem) {
+                        Ok(trait_did) => trait_did,
+                        Err(err) => cx.sess.fatal(&err[..])
```
Comment (eddyb) [inline src/librustc_typeck/check/method/suggest.rs]: Maybe this should fall back gracefully.

File: src/librustc_typeck/check/method/suggest.rs [hunk 2]
```diff
@@ -61,12 +64,44 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
...
+                    };
+
+                    let field_ty = ty::lookup_field_type(cx, did, field.id, substs);
+                    let field_ty_substs = Substs::new_trait(vec![fcx.inh.infcx.next_ty_var()],
```
Comment (eddyb) [inline src/librustc_typeck/check/method/suggest.rs]: Name could be misleading, the substitutions are for `FnOnce`, more specifically `vec![A]` in `FnOnce<A>`.

File: src/librustc_typeck/check/method/suggest.rs (hunk 1 above)
Comment (Nashenas88) [inline src/librustc_typeck/check/method/suggest.rs]: Sorry about that. I meant to ask about this case before submitting the PR, but I forgot. What should I do here?

File: src/librustc_typeck/check/method/suggest.rs (hunk 2 above)
Comment (Nashenas88) [inline src/librustc_typeck/check/method/suggest.rs]: If that's the case, I don't understand this as clearly as I thought. I'll follow up with you soon on IRC.

File: src/librustc_typeck/check/method/suggest.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc_typeck/check/method/suggest.rs]: Either display the message unconditionally, or only if the type is a function pointer or closure, or never display the message.

File: src/librustc_typeck/check/method/suggest.rs [hunk 3]
```diff
@@ -61,12 +64,62 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
...
+                                                       expr_string, item_ustring));
+                        }
+                    };
+
+                    macro_rules! span_did_you_mean {
+                        () => {
+                            cx.sess.span_note(span, &format!("did you mean to write `{0}.{1}`?",
+                                                             expr_string, item_ustring));
+                        }
+                    };
```
Comment (Nashenas88) [inline src/librustc_typeck/check/method/suggest.rs]: I forgot to ask before pushing these changes, is it alright to use a macro here? Is a closure preferred?

Comment (eddyb) [inline src/librustc_typeck/check/method/suggest.rs]: I'd say closures are preferred.

File: src/librustc_typeck/check/method/suggest.rs [hunk 4]
```diff
@@ -61,12 +64,62 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
...
+                            if selcx.evaluate_obligation(&obligation) {
+                                span_stored_function!();
+                            } else {
+                                span_did_you_mean!();
+                            }
+                        });
+                    } else {
+                        match field_ty.sty {
+                            // fallback to matching a closure or function pointer
+                            ty::TyClosure(_,_) | ty::TyBareFn(None,_) => span_stored_function!(),
```
Comment (eddyb) [inline src/librustc_typeck/check/method/suggest.rs]: This can just be `ty::TyClosure(..) | ty::TyBareFn(..)`, no need to be more specific than that.

**Comment by Nashenas88** [comment]
bump

**Comment by huonw** [comment]
I'm unfamiliar with the best ways to drive the type checker for stuff like this, maybe @nikomatsakis or @eddyb are better reviewers.

r? @eddyb

**Comment by Nashenas88** [comment]
@huonw, @eddyb has already reviewed this commit.

**Comment by huonw** [comment]
He can make the final determination then. :)

**Pull request merged**